implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiDemandPausingOp.java

```diff
@@ -0,0 +1,230 @@
+        processor = new PausableProcessor(subscriber);
+        if (!lateSubscription || !paused.get()) { // if late subscription is disabled, we can subscribe now.
+            subscribed.set(true);
+            upstream().subscribe(processor);
+        }
+    }
+
+    @Override
+    public boolean isPaused() {
+        return paused.get();
+    }
+
+    @Override
+    public void pause() {
+        paused.set(true);
+    }
+
+    @Override
+    public void resume() {
+        if (paused.compareAndSet(true, false)) {
+            PausableProcessor p = processor;
+            if (p != null) {
+                if (lateSubscription && subscribed.compareAndSet(false, true)) {
+                    upstream().subscribe(p);
+                }
+                p.resume();
+            }
+        }
+    }
+
+    @Override
+    public int bufferSize() {
+        PausableProcessor p = processor;
+        if (p != null) {
+            return p.queueSize();
+        }
+        return 0;
+    }
+
+    @Override
+    public boolean clearBuffer() {
+        if (paused.get()) {
+            PausableProcessor p = processor;
+            if (p != null) {
+                p.clearQueue();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private class PausableProcessor extends MultiOperatorProcessor<T, T> {
+
+        private final AtomicLong demand = new AtomicLong();
+        private final Queue<T> queue;
+        private final int maxBufferSize;
+
+        PausableProcessor(MultiSubscriber<? super T> downstream) {
+            super(downstream);
+            // Determine if we need a queue based on strategy and buffer size
+            if (backPressureStrategy == BackPressureStrategy.BUFFER) {
+                Supplier<Queue<T>> qs = bufferSize == 0 ? Queues.unbounded(Infrastructure.getMultiOverflowDefaultBufferSize())
+                        : Queues.get(bufferSize);
+                this.queue = qs.get();
+            } else {
+                this.queue = null;
+            }
+            this.maxBufferSize = bufferSize;
+        }
+
+        void resume() {
+            Flow.Subscription subscription = getUpstreamSubscription();
+            if (subscription == Subscriptions.CANCELLED) {
+                return;
+            }
+            // Drain any buffered items first
+            drainQueue();
+            long currentDemand = demand.get();
+            if (currentDemand > 0) {
+                Subscriptions.produced(demand, currentDemand);
+                subscription.request(currentDemand);
+            }
+        }
+
+        void drainQueue() {
+            if (queue == null) {
+                return;
+            }
+            Queue<T> qe = queue;
+            // Drain all buffered items - these were already requested from upstream
+            // so we don't need to check downstream demand here
+            while (!qe.isEmpty() && !paused.get()) {
```

Comment:
You can simply `poll` and check for `null`, it will save a `isEmpty()` call (which should not be costly, but...)